### PR TITLE
Implement block volumes with gluster-block

### PR DESCRIPTION
**Pull request by obnoxxx**

### CODE REVIEW

**Comment by obnoxxx** [comment]
Note: the first few unrelated commits are to be found in prs #790, #792, and  #791.
Can rebase once they are merged.

Needs more unit tests.

**Comment by lpabon** [comment]
This is going to be discussed at the meeting on [July 11](http://lists.gluster.org/pipermail/heketi-devel/2017-June/000154.html)

**Comment by MohamedAshiqrh** [comment]
@heketi/maintainers This patch is up for review. please do

heketi-cli block-volume <create/delete/info> --help 
heketi-cli cluster create --help
heketi-cli volume create --help

To get more info.

Also look at the changes in etc/heketi/heketi.json.

File: apps/glusterfs/app.go [hunk 1]
```diff
@@ -233,6 +247,25 @@ func (a *App) Upgrade(tx *bolt.Tx) error {
...
+	if "" != env {
+		a.conf.CreateBlockHostingVolumes, err = strconv.ParseBool(env)
+		if err != nil {
+			logger.LogError("Error: Parse bool in Create Block Hosting Volumes: %v", err)
+		}
+	}
+
+	env = os.Getenv("HEKETI_NEW_BLOCK_HOSTING_VOLUME_SIZE")
+	if "" != env {
+		a.conf.NewBlockHostingVolumeSize, err = strconv.Atoi(env)
```
Comment (lpabon) [inline apps/glusterfs/app.go]: Should there be a check here with the constants on the max/min volume size? It can probably be done in setBlockSettings()

File: apps/glusterfs/app.go [hunk 2]
```diff
@@ -256,6 +289,21 @@ func (a *App) setAdvSettings() {
 	}
 }
 
+func (a *App) setBlockSettings() {
```
Comment (lpabon) [inline apps/glusterfs/app.go]: Should this function return an error once it checks the values for correctness?

File: apps/glusterfs/app_block_volume.go [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+
+	if msg.Size < 1 {
+		http.Error(w, "Invalid volume size", http.StatusBadRequest)
+		logger.LogError("Invalid volume size")
+		return
+	}
+
+	if msg.Size > NewBlockHostingVolumeSize {
+		http.Error(w, "Default Block Hosting Volume size is less than block volume requested.", http.StatusBadRequest)
+		logger.LogError("Default Block Hosting Volume size is less than block volume requested.")
```
Comment (lpabon) [inline apps/glusterfs/app_block_volume.go]: To avoid copying the same string, replace as follows:

```go
err := logger.LogError("Some error")
http.Error(w, err.Error(), ...)
```

File: apps/glusterfs/app_block_volume.go [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+			http.Error(w, fmt.Sprintf("No clusters configured"), http.StatusBadRequest)
+			logger.LogError("No clusters configured")
+			return ErrNotFound
+		}
+
+		// Check the clusters requested are correct
+		for _, clusterid := range msg.Clusters {
+			_, err := NewClusterEntryFromId(tx, clusterid)
+			if err != nil {
+				http.Error(w, fmt.Sprintf("Cluster id %v not found", clusterid), http.StatusBadRequest)
```
Comment (lpabon) [inline apps/glusterfs/app_block_volume.go]: Same as above, but here it is simpler:

```go
err := logger.LogError("Clustert id %v not found", clusterid)
http.Error(w, err.Error(), ...)
```

File: apps/glusterfs/app_block_volume_test.go [hunk 5]
```diff
@@ -0,0 +1,565 @@
+//
```
Comment (lpabon) [inline apps/glusterfs/app_block_volume_test.go]: There should be tests added to check the cluster settings. If the cluster has Block:false, then this should return XXX according to the API

File: apps/glusterfs/app_volume.go [hunk 6]
```diff
@@ -246,8 +246,22 @@ func (a *App) VolumeDelete(w http.ResponseWriter, r *http.Request) {
...
+
+		if volume.Info.BlockInfo.BlockVolumes == nil {
+			return nil
+		}
+
+		if len(volume.Info.BlockInfo.BlockVolumes) == 0 {
+			return nil
+		}
 
+		err = fmt.Errorf("Cannot delete a block hosting volume containing block volumes")
```
Comment (lpabon) [inline apps/glusterfs/app_volume.go]: Use `err = logger.LogError(...)` so that it logs the error and returns it as an _error_ type

File: apps/glusterfs/app_volume.go [hunk 7]
```diff
@@ -246,8 +246,22 @@ func (a *App) VolumeDelete(w http.ResponseWriter, r *http.Request) {
 			return err
 		}
 
-		return nil
+		if !volume.Info.Block {
```
Comment (lpabon) [inline apps/glusterfs/app_volume.go]: This is strange. Can a BlockCreate point to a specific volume already created? (**We should have reviewed the API specification first**)

File: apps/glusterfs/block_volume_entry_create.go [hunk 8]
```diff
@@ -56,9 +56,16 @@ func (v *BlockVolumeEntry) createBlockVolumeRequest(db *bolt.DB,
 			return err
 		}
 
-		v.Info.BlockVolume.Hosts = bhvol.Info.Mount.GlusterFS.Hosts
-		v.Info.Hacount = len(v.Info.BlockVolume.Hosts)
+		if v.Info.Hacount > 0 {
+			for i := 0; i < v.Info.Hacount; i++ {
+				v.Info.BlockVolume.Hosts = append(v.Info.BlockVolume.Hosts, bhvol.Info.Mount.GlusterFS.Hosts[i])
```
Comment (humblec) [inline apps/glusterfs/block_volume_entry_create.go]: @MohamedAshiqrh iic, mount hosts are brick peers and NOT TSP peers. If yes, we are sending a wrong list. We should send TSP list instead of brick peers list.

**Comment by MohamedAshiqrh** [comment]
Rebased to the latest master

File: apps/glusterfs/app.go (hunk 2 above)
Comment (MohamedAshiqrh) [inline apps/glusterfs/app.go]: If values is not correct then this will fall back to default. I will add that correctness.

File: apps/glusterfs/app.go (hunk 1 above)
Comment (MohamedAshiqrh) [inline apps/glusterfs/app.go]: Added correctness. If given an invalid value it will fall back to default values.

File: apps/glusterfs/app_volume.go (hunk 7 above)
Comment (MohamedAshiqrh) [inline apps/glusterfs/app_volume.go]: So there are two volumes 

* Block Volumes
* Block Hosting Volumes(File volumes)

When a Block volume of 1 GB is requested, heketi creates a block hosting volume which will be of size 1TB(default value which is configurable). This block hosting volume will be used by gluster block to create 1 GB volume. The remaining size will be used for future requests, So BlockCreate will check if there are block hosting volume already present and then if there is space in them to allocate block volume requested. If not create a new block hosting volume. This approach is documented in https://github.com/gluster/gluster-kubernetes/pull/268.

File: apps/glusterfs/block_volume_entry_create.go (hunk 8 above)
Comment (MohamedAshiqrh) [inline apps/glusterfs/block_volume_entry_create.go]: This will be addressed by another PR which I am working on. The PR will make volfile-server the whole Trusted Storage Pool.

File: apps/glusterfs/app_volume.go (hunk 6 above)
Comment (MohamedAshiqrh) [inline apps/glusterfs/app_volume.go]: done

File: apps/glusterfs/app_block_volume.go (hunk 3 above)
Comment (MohamedAshiqrh) [inline apps/glusterfs/app_block_volume.go]: done

File: apps/glusterfs/app_block_volume.go (hunk 4 above)
Comment (MohamedAshiqrh) [inline apps/glusterfs/app_block_volume.go]: done

**Comment by MohamedAshiqrh** [comment]
@lpabon Have addressed all the comments except for test. Will add the test soon. Can you review the rest?

File: apps/glusterfs/app_volume.go (hunk 7 above)
Comment (lpabon) [inline apps/glusterfs/app_volume.go]: Thanks @MohamedAshiqrh , I would like to see a design document for this feature as part of this patch. We can add it to doc/blockvolume.md or something. This doc should highlight (not to deeply) how Heketi is supporting this and how the internals work.

Comment (obnoxxx) [inline apps/glusterfs/app_volume.go]: @lpabon this is all in the end-to-end design document proposed under gluster-kubernetes: https://github.com/gluster/gluster-kubernetes/pull/268

Comment (obnoxxx) [inline apps/glusterfs/app_volume.go]: @lpabon we can at some point copy the contents here, but I wanted to propose it in one central place and not scatter or duplicate it for a start.

Comment (obnoxxx) [inline apps/glusterfs/app_volume.go]: @lpabon the latest copy of the design doc in the PR can be seen under https://github.com/obnoxxx/gluster-kubernetes/blob/gluster-block-design/docs/design/gluster-block-provisioning.md

Comment (lpabon) [inline apps/glusterfs/app_volume.go]: We need a document specific to the API and design here in this repo as part of this engineering effort. The overall project is in gluster-kubernetes.

File: apps/glusterfs/block_volume_entry_create.go [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+package glusterfs
+
+import (
+	"github.com/boltdb/bolt"
+	"github.com/heketi/heketi/executors"
+	"github.com/lpabon/godbc"
+)
+
+func (v *BlockVolumeEntry) createBlockVolume(db *bolt.DB,
```
Comment (lpabon) [inline apps/glusterfs/block_volume_entry_create.go]: Should add comment that this function requires that a volume be created and exist for it to create a block device.

File: apps/glusterfs/block_volume_entry_create.go [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+func (v *BlockVolumeEntry) createBlockVolumeRequest(db *bolt.DB,
+	executor executors.Executor,
+	blockHostingVolumeId string) (*executors.BlockVolumeRequest, string, error) {
+	godbc.Require(db != nil)
+	godbc.Require(blockHostingVolumeId != "")
+
+	var blockHostingVolumeName string
+
+	err := db.View(func(tx *bolt.Tx) error {
+		logger.Debug("Looking for block hosting volume %v", blockHostingVolumeId)
```
Comment (lpabon) [inline apps/glusterfs/block_volume_entry_create.go]: Looking or getting info from?

File: apps/glusterfs/block_volume_entry_create.go [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+		bhvol, err := NewVolumeEntryFromId(tx, blockHostingVolumeId)
+		if err != nil {
+			return err
+		}
+
+		if v.Info.Hacount > 0 {
+			for i := 0; i < v.Info.Hacount; i++ {
```
Comment (lpabon) [inline apps/glusterfs/block_volume_entry_create.go]: What happens if `Hacount` is larger than len(bhvol.Info.Mount.GlusterFS.Hosts)? This would panic. Before the `for` loop a decision needs to be made. Either we ignore the Hacount and set it to a value equal to the len(bhvol.Info.Mount.GlusterFS.Hosts) or we return an error because the Hacount cannot be satisfied.

File: apps/glusterfs/block_volume_entry_create.go [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+		blockHostingVolumeName = bhvol.Info.Name
+
+		return nil
+	})
+	if err != nil {
+		logger.Err(err)
+		return nil, "", err
+	}
+
+	executorhost, err := GetVerifiedManageHostname(db, executor, v.Info.Cluster)
```
Comment (lpabon) [inline apps/glusterfs/block_volume_entry_create.go]: Comment. What is this doing? Why do we need this? Simple one line comments are nice :-)

**Comment by lpabon** [comment]
retest this please

**Comment by MohamedAshiqrh** [comment]
Rebase to have funtional test fixes.

File: apps/glusterfs/block_volume_entry_create.go (hunk 12 above)
Comment (MohamedAshiqrh) [inline apps/glusterfs/block_volume_entry_create.go]: Done.

File: apps/glusterfs/block_volume_entry_create.go (hunk 11 above)
Comment (MohamedAshiqrh) [inline apps/glusterfs/block_volume_entry_create.go]: Fixed to set hacount to the len(bhvol.Info.Mount.GlusterFS.Hosts).

File: apps/glusterfs/block_volume_entry_create.go (hunk 10 above)
Comment (MohamedAshiqrh) [inline apps/glusterfs/block_volume_entry_create.go]: Done

File: apps/glusterfs/block_volume_entry_create.go [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+	blockVolumeInfo, err := executor.BlockVolumeCreate(host, vr)
+	if err != nil {
+		return err
+	}
+
+	v.Info.BlockVolume.Iqn = blockVolumeInfo.Iqn
+	v.Info.BlockVolume.Hosts = blockVolumeInfo.BlockHosts
+	v.Info.BlockVolume.Lun = 0
+	v.Info.BlockVolume.Username = blockVolumeInfo.Username
+	v.Info.BlockVolume.Password = blockVolumeInfo.Password
```
Comment (MohamedAshiqrh) [inline apps/glusterfs/block_volume_entry_create.go]: I am not sure what does clear text mean. Yes, this will be saved in the db. These are parameters returned from the gluster-block for the volume create.

File: apps/glusterfs/block_volume_entry_create.go [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+	godbc.Require(db != nil)
+	godbc.Require(blockHostingVolumeId != "")
+
+	vr, host, err := v.createBlockVolumeRequest(db, executor,
+		blockHostingVolumeId)
+	if err != nil {
+		return err
+	}
+
+	blockVolumeInfo, err := executor.BlockVolumeCreate(host, vr)
```
Comment (MohamedAshiqrh) [inline apps/glusterfs/block_volume_entry_create.go]: done

File: apps/glusterfs/block_volume_entry_create.go (hunk 9 above)
Comment (MohamedAshiqrh) [inline apps/glusterfs/block_volume_entry_create.go]: Done

File: apps/glusterfs/block_volume_entry_create.go [hunk 15]
```diff
@@ -0,0 +1,96 @@
+//
+// Copyright (c) 2015 The heketi Authors
```
Comment (MohamedAshiqrh) [inline apps/glusterfs/block_volume_entry_create.go]: Done

File: apps/glusterfs/block_settings.go [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+// cases as published by the Free Software Foundation.
+//
+
+package glusterfs
+
+var (
+	// Default block settings
+	CreateBlockHostingVolumes = false
+	// Default 1 TB
+	NewBlockHostingVolumeSize = 1024
```
Comment (MohamedAshiqrh) [inline apps/glusterfs/block_settings.go]: Done

File: apps/glusterfs/app_config.go [hunk 17]
```diff
@@ -30,6 +30,10 @@ type GlusterFSConfig struct {
 	BrickMaxSize int `json:"brick_max_size_gb"`
 	BrickMinSize int `json:"brick_min_size_gb"`
 	BrickMaxNum  int `json:"max_bricks_per_volume"`
+
+	//block settings
+	CreateBlockHostingVolumes bool `json:"auto_create_block_hosting_volume"`
+	NewBlockHostingVolumeSize int  `json:"new_block_hosting_volume_size"`
```
Comment (MohamedAshiqrh) [inline apps/glusterfs/app_config.go]: Done

File: apps/glusterfs/app_block_volume.go [hunk 18]
```diff
@@ -0,0 +1,196 @@
...
+	if err := json.NewEncoder(w).Encode(info); err != nil {
+		panic(err)
+	}
+}
+
+func (a *App) BlockVolumeDelete(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	id := vars["id"]
+
+	var volume *BlockVolumeEntry
```
Comment (MohamedAshiqrh) [inline apps/glusterfs/app_block_volume.go]: Done

File: apps/glusterfs/app_block_volume.go [hunk 19]
```diff
@@ -0,0 +1,196 @@
...
+		return
+	}
+
+	if msg.Size < 1 {
+		http.Error(w, "Invalid volume size", http.StatusBadRequest)
+		logger.LogError("Invalid volume size")
+		return
+	}
+
+	if msg.Size > NewBlockHostingVolumeSize {
```
Comment (MohamedAshiqrh) [inline apps/glusterfs/app_block_volume.go]: @lpabon The NewBlockHostingVolumeSize is a configurable value, if blockvolume has to be created of bigger size then they can change the value and create the blockvolume. 

@obnoxxx Can you share your thoughts too.

**Comment by MohamedAshiqrh** [comment]
@lpabon addressed all the comments.

File: executors/sshexec/block_volume.go [hunk 20]
```diff
@@ -0,0 +1,130 @@
...
+	}
+
+	var auth_set string
+	if volume.Auth {
+		auth_set = "enable"
+	} else {
+		auth_set = "disable"
+	}
+
+	cmd := fmt.Sprintf("gluster-block create %v/%v  ha %v auth %v  %v %vG --json",
```
Comment (humblec) [inline executors/sshexec/block_volume.go]: @MohamedAshiqrh I think, gluster-block expects `GiB` instead of "G".

Comment (MohamedAshiqrh) [inline executors/sshexec/block_volume.go]: @humblec https://github.com/gluster/gluster-block/blob/94c79620db4451bf804d6ab631c9ca59759dbc21/utils/common.c#L78

I referred from there.

Comment (MohamedAshiqrh) [inline executors/sshexec/block_volume.go]: Which means both will work. Just wants a capital G to start.

**Comment by MohamedAshiqrh** [comment]
@lpabon I have addressed almost all the comments(except for adding a test and create a doc from the existing [doc](https://github.com/obnoxxx/gluster-kubernetes/blob/gluster-block-design/docs/design/gluster-block-provisioning.md) with only specifics of heketi API). 

Can you review the rest?

File: apps/glusterfs/app_volume.go (hunk 7 above)
Comment (obnoxxx) [inline apps/glusterfs/app_volume.go]: The details are currenlty in the other repo, because I wanted to have a single document describing it all and wanted to avoid duplication of document and effort. We can rip the heketi part out of the other repo and put it here, but I somehow had hoped that it would not be asking too much to look at the gluster-kubernetes repo which is associated with heketi anyways. ;-)

**Comment by obnoxxx** [comment]
@MohamedAshiqrh Thanks a lot for all the updates!

**Comment by obnoxxx** [comment]
retest this please @centos-ci

**Comment by MohamedAshiqrh** [comment]
rebased.

**Comment by obnoxxx** [comment]
retest this please @centos-ci

**Comment by obnoxxx** [comment]
couldn't fully test this locally since functional tests don't work for me. hence pushing this in

**Comment by obnoxxx** [comment]
retest this please @centos-ci

**Comment by obnoxxx** [comment]
whoo, ci functional selftest passed :-)

**Comment by MohamedAshiqrh** [comment]
yay...! CI worked @obnoxxx nice work

File: executors/sshexec/block_volume.go [hunk 21]
```diff
@@ -0,0 +1,131 @@
...
+		volume.GlusterVolumeName, volume.Name, volume.Hacount, auth_set, strings.Join(volume.BlockHosts, ","), volume.Size)
```
Comment (humblec) [inline executors/sshexec/block_volume.go]: @MohamedAshiqrh we also need to enable the gluster-block-profile on the hosting FS volume at this time.

Comment (humblec) [reply executors/sshexec/block_volume.go]: @obnoxxx `gluster volume set <vol1> group gluster-block and restart the volume` is required for each block hosting volume.

File: executors/sshexec/block_volume.go (hunk 20 above)
Comment (raghavendra-talur) [reply executors/sshexec/block_volume.go]: We need to call gluster-block with "prealloc full"

File: executors/sshexec/block_volume.go (hunk 21 above)
Comment (obnoxxx) [reply executors/sshexec/block_volume.go]: thanks, @humblec will work this in

File: executors/sshexec/block_volume.go [hunk 22]
```diff
@@ -42,7 +42,7 @@ func (s *SshExecutor) BlockVolumeCreate(host string,
 		auth_set = "disable"
 	}
 
-	cmd := fmt.Sprintf("gluster-block create %v/%v  ha %v auth %v  %v %vG --json",
+	cmd := fmt.Sprintf("gluster-block create %v/%v  prealloc full ha %v auth %v  %v %vG --json",
```
Comment (humblec) [inline executors/sshexec/block_volume.go]: @obnoxxx  @raghavendra-talur this `prealloc full` has broken the command line.  Have you tested it ?

Comment (humblec) [reply executors/sshexec/block_volume.go]: @obnoxxx jfyi  ^^^^

**Comment by obnoxxx** [comment]
@raghavendra-talur i rebased + force-pushed. your conflicts are resolved.

**Comment by humblec** [comment]
Thanks @obnoxxx and @raghavendra-talur 👍

**Comment by obnoxxx** [comment]
superseded by polished version #853
